numba/typed/typedobjectutils.py

```diff
@@ -197,3 +197,80 @@ def build_wrapper(fn):
     builder.ret(context.get_constant(types.int32, -1))
 
     return equal_fn
+
+
+def _get_primitive_equal(context, module, datamodel, func_name_suffix):
+    # construct equal_fn
+    data_ptr_ty = ir.IntType(8).as_pointer()
+    equal_fnty = ir.FunctionType(ir.IntType(32), [data_ptr_ty, data_ptr_ty])
+    equal_fn = cgutils.get_or_insert_function(
+        module,
+        equal_fnty,
+        name=f".numba_{context.fndesc.mangled_name}.{func_name_suffix}_equal",
+    )
+
+    # populate the body of equal_fn
+    builder = ir.IRBuilder(equal_fn.append_basic_block())
+    lhs_ptr = builder.bitcast(equal_fn.args[0], datamodel.be_type.as_pointer())
+    lhs = builder.load(lhs_ptr)
+    rhs_ptr = builder.bitcast(equal_fn.args[1], datamodel.be_type.as_pointer())
+    rhs = builder.load(rhs_ptr)
+
+    if datamodel.fe_type.signed:
+        res = builder.icmp_signed('==', lhs, rhs)
+    else:
+        res = builder.icmp_unsigned('==', lhs, rhs)
+
+    with builder.if_else(res) as (then, orelse):
+        with then:
+            builder.ret(context.get_constant_generic(builder, types.int32, 1))
+        with orelse:
+            builder.ret(context.get_constant_generic(builder, types.int32, 0))
+
+    builder.ret(context.get_constant_generic(builder, types.int32, -1))
+
+    return equal_fn
```

Comment:
You can simplify the implementation like this:

```python
def _get_primitive_equal(context, module, datamodel, container_element_type):
    # construct equal_fn
    i8 = ir.IntType(8)
    i8p = i8.as_pointer()
    i32 = ir.IntType(32)

    equal_fnty = ir.FunctionType(i32, [i8p, i8p])
    equal_fn = cgutils.get_or_insert_function(
        module,
        equal_fnty,
        name=f".numba_{context.fndesc.mangled_name}.{container_element_type}_equal",
    )

    # populate the body of equal_fn
    builder = ir.IRBuilder(equal_fn.append_basic_block())
    lhs_ptr = builder.bitcast(equal_fn.args[0], datamodel.be_type.as_pointer())
    lhs = builder.load(lhs_ptr)
    rhs_ptr = builder.bitcast(equal_fn.args[1], datamodel.be_type.as_pointer())
    rhs = builder.load(rhs_ptr)

    if datamodel.fe_type.signed:
        res = builder.icmp_signed('==', lhs, rhs)
    else:
        res = builder.icmp_unsigned('==', lhs, rhs)
    builder.ret(builder.zext(res, i32))
    return equal_fn
```

Reply:
sure, simplified this function. especially use `zext` to avoid explicit if-branch.